mshadow/half.h

```diff
@@ -159,12 +165,18 @@ class MSHADOW_ALIGNED(2) half_t {
     uint32_t ui;
```

Comment:
I have a stupid question: what does si and ui stand for?

Reply:
The 'Bits' union:
```
union Bits {
    float f;
    int32_t si;
    uint32_t ui;
};
```
... has 3 'views' of the underlying 32-bits of storage: float (.f), signed integer (.si) and unsigned integer (.ui).  The need for both si and ui here is not totally obvious- probably enables fine-tuning the implementation for a slight perf advantage or perhaps to avoid compiler warnings when combining signed and unsigned values.

Also, if it helps the review any, the approach to rounding denorms is identical in this PR to the recently merged numpy code (made in response to an issue I raised): https://github.com/numpy/numpy/pull/12845